src/shared/sleep-config.c

```diff
@@ -102,6 +107,174 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
+
+        if (battery_capacity < 0 || battery_capacity > 100)
+               return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Invalid battery capacity");
+
+        return battery_capacity;
+}
+
+/* Read system and battery identifier from specific location and generate hash of it */
+static int get_system_battery_identifier_hash(uint8_t ret[static SHA256_DIGEST_SIZE]) {
+        _cleanup_free_ char *bat_manufacturer = NULL, *bat_model_name = NULL, *bat_serial_number = NULL;
+        sd_id128_t machine_id, product_id;
+        struct sha256_ctx hash;
+        int r;
+
+        sha256_init_ctx(&hash);
+        r = read_one_line_file("/sys/class/power_supply/BAT0/manufacturer", &bat_manufacturer);
+        if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/manufacturer is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/manufacturer: %m");
+        else
+               sha256_process_bytes(bat_manufacturer, strlen(bat_manufacturer) + 1, &hash);
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/model_name", &bat_model_name);
+        if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/model_name is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/model_name: %m");
+        else
+               sha256_process_bytes(bat_model_name, strlen(bat_model_name) + 1, &hash);
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/serial_number", &bat_serial_number);
+        if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/serial_number is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/serial_number: %m");
+        else
+               sha256_process_bytes(bat_serial_number, strlen(bat_serial_number) + 1, &hash);
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r == -ENOENT)
+               log_debug_errno(r, "machine ID is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to get machine ID: %m");
+        else
+               sha256_process_bytes(&machine_id, sizeof(machine_id), &hash);
+
+        r = id128_get_product(&product_id);
+        if (r == -ENOENT)
+               log_debug_errno(r, "product_id does not exist: %m");
+        else if (r < 0)
+                return log_debug_errno(r, "Failed to get product ID: %m");
+        else
+                sha256_process_bytes(&product_id, sizeof(product_id), &hash);
+
+        sha256_finish_ctx(&hash, ret);
+
+        return 0;
+}
+
+/* battery percentage discharge rate per hour is in range 1-199 then return success */
+static bool battery_discharge_rate_is_valid(int battery_discharge_rate) {
+        return battery_discharge_rate > 0 && battery_discharge_rate < 200;
+}
```

Comment:
The upper bound `200` is not correct.
E.g. if the battery is discharged 50% in 1min, then the rate is 3000.
Please use `uint64_t` for the value, as the theoretical maximum is `1 * USEC_PER_HOUR`.

Reply:
You mean change the upper bound or change the data type for battery_discharge_rate?

Reply:
Please change the type and accept any positive values.

Reply:
if we do that and "E.g. if the battery is discharged 50% in 1min, then the rate is 3000." any such case is missed, we might end up suspending system instead when it should be hibernated

Reply:
Yeah I don't think there's any use in going outside the current limits, the point is to suspend for a reasonable amount of time because it's a nicer user experience than hibernating. Suspending for a few seconds and then hibernating is pointless, in those cases we should just go to hibernation directly.

Reply:
But the function is a kind of library-like function. Why not to accept general value? Such a meaningless value should be excluded in the main loop.

Reply:
If we store such very large value, then the condition below in the loop will prevent the _next_ suspend.
```c
r = get_battery_discharge_rate();
if (last_capacity * 2 <= r)
        break
```
That's useful and I guess this is exactly the purpose of the value saved in the file under /var.
Otherwise, systems whose battery discharge so fast will fail to suspend every time, no?

Reply:
Here we are handling case till the point when battery can be completly drained in 30 min atleast to save system state and also don't let complete battery drain out. If any system battery will discharge in less than 30 min, we may loose the system state if we keep it suspended for few minutes and then try hibernating. For eg :
Discharge rate is 50% per 1 min. Then in 2 min system would die. If we put such a system to suspend in initial loop, and during wakeup the system took more than 1 min to restore state, you know like process delay after hardware resumes, we may loose the system state as system might shut down in that short while due to low battery.
You have got the motivation right. It is to save such systems from getting critical and then die which is only possible if we save state in hibernation long enough before. And that is the reason to hold 30 min buffer and to make it same across all sorts of situation, the scenario where the system might die in less than 30 min, it should be hibernated.

Reply:
Still I do not agree.
Let's consider the following senario.
1. A laptop has a fully charged battery that discharges 10% per 1min (the previous example 50% per 1min is hopeless, hence I relaxed that :-D).
2. Assume that initially we do not have the saved discharge rate file.
3. So, the laptop enters suspend as `battery_is_low() == 0`.
4. Luckily, the user of the laptop wake the laptop up after 3min.
5. We calculate and obtain the discharge rate (10% per 1min = 6000% per 1hour). The current code refuse the value to save. What I wantt to emphasize here is that we should save the rate anyway.

- If the discharge rate **is not** saved, when the user try to suspend the laptop next time, we allow that. And if the laptop is not woken up by the user within a short time, data is lost...
- If the discharge rate **is** saved, on the next trial of suspend, we refuse that, and the laptop will be hibernated. Everything fine, no??

Reply:
Hmm. That makes sense but i think enabling ACPI _BTP will handle that. That is the next task in this queue.

Reply:
> Still I do not agree. Let's consider the following senario.
> 
> 1. A laptop has a fully charged battery that discharges 10% per 1min (the previous example 50% per 1min is hopeless, hence I relaxed that :-D).
> 2. Assume that initially we do not have the saved discharge rate file.
> 3. So, the laptop enters suspend as `battery_is_low() == 0`.
> 4. Luckily, the user of the laptop wake the laptop up after 3min.
> 5. We calculate and obtain the discharge rate (10% per 1min = 6000% per 1hour). The current code refuse the value to save. What I wantt to emphasize here is that we should save the rate anyway.
> 
> * If the discharge rate **is not** saved, when the user try to suspend the laptop next time, we allow that. And if the laptop is not woken up by the user within a short time, data is lost...
> * If the discharge rate **is** saved, on the next trial of suspend, we refuse that, and the laptop will be hibernated. Everything fine, no??

That seems like a very very unlikely corner case. If your battery lasts 10 minutes while on suspend, you are not going to suspend in the first place... it wouldn't make sense to do so. Nonetheless, the current situation is that we suspend for 1hr and that's it. So for that corner case there's really no difference. It's unrealistic, and if it happens behaviour is the same as it is now.

Reply:
Still I do not agree with that point. But the condition can be easily dropped later. Let's merge this as is.